pkg/cmd/admin/node/node_options.go

```diff
@@ -0,0 +1,155 @@
+package node
+
+import (
+	"errors"
+	"io"
+
+	"github.com/spf13/cobra"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/meta"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl"
+	kcmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/util/errors"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+type NodeOptions struct {
+	DefaultNamespace string
+	Kclient          *client.Client
+	Writer           io.Writer
+
+	Mapper            meta.RESTMapper
+	Typer             runtime.ObjectTyper
+	RESTClientFactory func(mapping *meta.RESTMapping) (resource.RESTClient, error)
+	Printer           func(mapping *meta.RESTMapping, noHeaders bool) (kubectl.ResourcePrinter, error)
+
+	CmdPrinter       kubectl.ResourcePrinter
+	CmdPrinterOutput bool
+
+	NodeNames []string
+
+	// Common optional params
+	Selector    string
+	PodSelector string
+}
+
+func (n *NodeOptions) Complete(f *clientcmd.Factory, c *cobra.Command, args []string, out io.Writer) error {
+	defaultNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	_, kc, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	cmdPrinter, output, err := kcmdutil.PrinterForCommand(c)
+	if err != nil {
+		return err
+	}
+	mapper, typer := f.Object()
+
+	n.DefaultNamespace = defaultNamespace
+	n.Kclient = kc
+	n.Writer = out
+	n.Mapper = mapper
+	n.Typer = typer
+	n.RESTClientFactory = f.Factory.RESTClient
+	n.Printer = f.Printer
+	n.NodeNames = []string{}
+	n.CmdPrinter = cmdPrinter
+	n.CmdPrinterOutput = false
+
+	if output {
+		n.CmdPrinterOutput = true
+	}
+	if len(args) != 0 {
+		n.NodeNames = append(n.NodeNames, args...)
+	}
+	return nil
+}
+
+func (n *NodeOptions) Validate() error {
+	errList := []error{}
+	if len(n.Selector) > 0 {
+		if _, err := labels.Parse(n.Selector); err != nil {
+			errList = append(errList, errors.New("--selector=<node_selector> must be a valid label selector"))
+		}
+		if len(n.NodeNames) != 0 {
+			errList = append(errList, errors.New("either specify --selector=<node_selector> or nodes but not both"))
+		}
+	}
+
+	if len(n.PodSelector) > 0 {
+		if _, err := labels.Parse(n.PodSelector); err != nil {
+			errList = append(errList, errors.New("--pod-selector=<pod_selector> must be a valid label selector"))
+		}
+	}
+	return kerrors.NewAggregate(errList)
+}
+
+func (n *NodeOptions) GetNodeResource() (*resource.Result, error) {
```

Comment:
Based on the usage I'm seeing, you actually want this method to return `[]*kapi.Node, error`.  No other part of the result seems to be used.  You could scrub the results here.  That will make the rest of the code easier to read by avoiding the visitor.